### PR TITLE
Revert "Fix dockerfile parser with empty line after escape"

### LINKED ISSUE #29005: Parser in docker 1.13 breaks backwards compatibility for comments in middle of RUN statement
**Description**
Previously (Docker 1.6+) one could document long `RUN` lines with comments interspersed. https://github.com/docker/docker/pull/24725 broke this.
```Dockerfile
FROM alpine
RUN echo \
# comment
hi
```
**Steps to reproduce the issue:**
1. build the above Dockerfile on 1.13-rc2

**Describe the results you received:**
`Error response from daemon: Unknown instruction: HI`

**Describe the results you expected:**
 Docker build the above Dockerfile on 1.6 through 1.12 (thanks @tianon for testing all of them)
`Successfully built xxx`

**Additional information you deem important (e.g. issue happens only occasionally):**
Quick list of some official-images broken by this change:

- [arangodb:2.5.5](https://github.com/arangodb/arangodb-docker/tree/239629aa99849ff2447269318f668085eb2c5168/jessie/2.5.5) from June 2015
  - and all though [arangodb:3.1.2](https://github.com/arangodb/arangodb-docker/blob/239629aa99849ff2447269318f668085eb2c5168/jessie/3.1.2/Dockerfile)
- [bash](https://github.com/tianon/docker-bash/blob/d3283a47fbb92419b7b6fc1fccd30b95d75e9ab6/4.4/Dockerfile)
- [cassandra](https://github.com/docker-library/cassandra/blob/d3a91560b21e73994235a72a4c3153e775e1654d/Dockerfile.template)
- [elasticsearch](https://github.com/docker-library/elasticsearch/blob/4ddeedae52db2a71a2022fe268f0454eb0e913e4/Dockerfile.template)
- [golang:windowsservercore](https://github.com/docker-library/golang/blob/4fd5df86eea53623b1009b3621b40a97f9f359e5/1.7/windows/windowsservercore/Dockerfile) and [nanoserver](https://github.com/docker-library/golang/tree/4fd5df86eea53623b1009b3621b40a97f9f359e5/1.7/windows/nanoserver)
- [httpd](https://github.com/docker-library/httpd/blob/3d40709ef614596b81f72e79fe3fca7358d2eeef/2.4/Dockerfile)
- [irssi](https://github.com/jessfraz/irssi/blob/e3c492d1c852de8a9393979a1989d19ac15821f6/debian/Dockerfile)
- [julia](https://github.com/docker-library/julia/blob/7e52c3cd99d1b39028bf4b36da493ec47cb643c2/Dockerfile)
- [kibana](https://github.com/docker-library/kibana/blob/960d3f8e41828a75585dbabc528d5e226b1adc3c/5.0/Dockerfile)
- [logstash](https://github.com/docker-library/logstash/blob/79bc3b0c32cb559654228ca79cc9ed4b0f1b158c/Dockerfile.template)
- [mariadb](https://github.com/docker-library/mariadb/blob/92c8ff1abdacd64359deb9a3de91c0dab6115739/Dockerfile.template)
- [mongo:windowsservercore](https://github.com/docker-library/mongo/blob/5f4bcf4bec163ef05b4fc67d5c92762989dbde06/3.4/windows/windowsservercore/Dockerfile)
- [mysql:5.5](https://github.com/docker-library/mysql/blob/400ca81ac2c7d0cc0a599c8b2f73fc0d0e731fb6/5.5/Dockerfile)

**Output of `docker version`:**
(working version)
```console
Client:
 Version:      1.12.3
 API version:  1.24
 Go version:   go1.7.3
 Git commit:   6b644ec
 Built:        
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.3
 API version:  1.24
 Go version:   go1.7.3
 Git commit:   6b644ec
 Built:        
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**
(working version)
```
Containers: 6
 Running: 4
 Paused: 0
 Stopped: 2
Images: 5715
Server Version: 1.12.3
Storage Driver: overlay
 Backing Filesystem: extfs
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: null overlay bridge host
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: seccomp
Kernel Version: 4.8.8-gentoo
Operating System: Gentoo/Linux
OSType: linux
Architecture: x86_64
CPUs: 8
Total Memory: 30.87 GiB
Name: isengard
ID: UOCM:3F65:5FZC:6H5L:W3HY:34G4:A5XZ:SUOV:S2D4:XQTO:4KGA:6XSE
Docker Root Dir: /mnt/spare/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Insecure Registries:
 127.0.0.0/8
```

### PR DESCRIPTION
**Pull request by vdemeester**
This reverts commit 3e1b539e8d0ed4abf695b0a8c42346fba6d5a6b0, in order to fix #29005 but reverting #24725.

The basic idea is the following :

* Reverting that altogether on 1.13 to not break people
* For 1.14, add it back but as a warning first and with a deprecation notice.
* When it's time, no more warning

/ping @yongtang can you work on that for 1.14 ? (this might close #29006 too)

/cc @tiborvass @thaJeztah @cpuguy83 @vieux @duglin @tianon @yosifkit

🐸

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

### CODE REVIEW

**Comment by runcom** [comment]
Sgtm

File: integration-cli/docker_cli_build_test.go [hunk 1]
```diff
@@ -7140,190 +7140,3 @@ func (s *DockerSuite) TestBuildNetContainer(c *check.C) {
...
-#echo env-load
-RUN echo vegeta
-`, true)
-	c.Assert(err, checker.IsNil)
-	c.Assert(out, checker.Contains, "Step 1/3 : FROM busybox")
-	c.Assert(out, checker.Contains, "Step 2/3 : RUN echo grafana &&     echo raintank")
-	c.Assert(out, checker.Contains, "Step 3/3 : RUN echo vegeta")
-}
-
-func (s *DockerSuite) TestBuildSquashParent(c *check.C) {
```
Comment (thaJeztah) [inline integration-cli/docker_cli_build_test.go]: Don't think this one must be removed

File: integration-cli/docker_cli_build_test.go [hunk 2]
```diff
@@ -7140,190 +7140,3 @@ func (s *DockerSuite) TestBuildNetContainer(c *check.C) {
...
-	splitOrigHistory := strings.Split(strings.TrimSpace(origHistory), "\n")
-	splitTestHistory := strings.Split(strings.TrimSpace(testHistory), "\n")
-	c.Assert(len(splitTestHistory), checker.Equals, len(splitOrigHistory)+1)
-
-	out, err = inspectImage(id, "len .RootFS.Layers")
-	c.Assert(err, checker.IsNil)
-	c.Assert(strings.TrimSpace(out), checker.Equals, "3")
-}
-
-func (s *DockerSuite) TestBuildContChar(c *check.C) {
```
Comment (thaJeztah) [inline integration-cli/docker_cli_build_test.go]: Don't think this one must be removed

File: integration-cli/docker_cli_build_test.go [hunk 3]
```diff
@@ -7140,190 +7140,3 @@ func (s *DockerSuite) TestBuildNetContainer(c *check.C) {
...
-		 RUN echo hi \\\`, true)
-	c.Assert(err, checker.IsNil)
-	c.Assert(out, checker.Contains, "Step 1/2 : FROM busybox")
-	c.Assert(out, checker.Contains, "Step 2/2 : RUN echo hi \\\\\n")
-}
-
-// TestBuildOpaqueDirectory tests that a build succeeds which
-// creates opaque directories.
-// See https://github.com/docker/docker/issues/25244
-func (s *DockerSuite) TestBuildOpaqueDirectory(c *check.C) {
```
Comment (thaJeztah) [inline integration-cli/docker_cli_build_test.go]: Same

Actually, looks like only the first one has to be removed; https://github.com/docker/docker/commit/3e1b539e8d0ed4abf695b0a8c42346fba6d5a6b0

**Review by thaJeztah — CHANGES_REQUESTED** [review]

**Comment by duglin** [comment]
Should we start the deprecation process in this release?

**Comment by thaJeztah** [comment]
@duglin I think that was option `C`, and not everyone was sure about that. Could be done in a follow-up for discussion I guess

**Comment by vdemeester** [comment]
@duglin We're getting pretty close to the release. I would tend to prefer start the depreciation process in the next release to not overdo this release, but as @thaJeztah says, could be done in a follow-up.

**Review by thaJeztah — APPROVED** [review]
LGTM

**Comment by thaJeztah** [comment]
oh, we should revert the changelog as well, and reopen https://github.com/docker/docker/issues/24693 after this was merged

**Comment by yongtang** [comment]
Thanks @vdemeester for the PR. I will work on the follow up PR for deprecation/warning based on the discussion in #29005.

**Comment by thaJeztah** [comment]
Thanks @yongtang, sorry for having this PR in parallel to yours, we wanted to have it ready for the next RC :heart:

**Comment by tiborvass** [comment]
LGTM

**Pull request merged**